### PR TITLE
ENH: Add default convex conjugate, closes #570

### LINKED ISSUE #570: Add a default convex conjugate
Currently in the functional class, `convex_conj` raises `NotImplementedError`. This is problematic for proximals where some algebra exists.

Consider if I implement a functional with a proximal

```python
class MyFancyFunctional(Functional):
    ...
    @property
    def proximal(self):
        return MyProximal
```

and I want to use this with a solver that takes the proximal of its convex conjugate. Here I would expect this to work due to the Moreau identity

```python
f = MyFancyFunctional(...)
my_solver(f.convex_conj.proximal)
```

But right now I simply get `NotImplementedError`.

We should implement a class, like `DefaultConvexConjugate`, which does not implement `_call`, but which implements `proximal`.

### PR DESCRIPTION
**Pull request by adler-j**
Cherry picked from #587, should be simple to review.

Things that need to be done:

* Add a test

Things that could be done:

* Add some intelligence so that this can only be created if it is useful (i.e. if the functional has a proximal). This could be nice in some cases I guess, but in others it could also give hard to follow logic.
* Perhaps add some logic to evaluate the functional in some sense (perhaps solve the optimization problem, somehow). Not sure at all if this is a good idea.

### CODE REVIEW

File: test/solvers/functional/functional_test.py [hunk 1]
```diff
@@ -280,11 +280,9 @@ def test_functional_sum(space):
                             func2.gradient(x).inner(p)),
                         places=PLACES)
 
-    # Test that prox and convex conjugate is not known
+    # Test that prox is not known
     with pytest.raises(NotImplementedError):
         (func1 + func2).proximal
-    with pytest.raises(NotImplementedError):
```
Comment (aringh) [inline test/solvers/functional/functional_test.py]: The convex conjugate should not be known for a sum. Wouldn't you still like to check something here, like that `(func1 + func2).convex_conj` is a `FunctionalDefaultConvexConjugate` or that  `(func1 + func2).convex_conj(x)` raises `NotImplemetedError`? Or are those kind of test just redundant?

File: odl/solvers/functional/functional.py [hunk 2]
```diff
@@ -500,7 +501,7 @@ def __init__(self, func, scalar):
             raise TypeError('`func` {!r} is not a `Functional` instance'
                             ''.format(func))
 
-        scalar = func.range.element(scalar)
+        scalar = func.domain.field.element(scalar)
```
Comment (aringh) [inline odl/solvers/functional/functional.py]: This should be the same, since `func` is checked to be a `functional` on the line above, and for `functional` we have `range = domain.field`.

Comment (kohr-h) [inline odl/solvers/functional/functional.py]: Currently yes, but once we got the math straight, we might in fact want allow spaces over the complex numbers as `domain`, and in that case, we want to cast to the field of the domain, not the range (which we want to keep as "always real", I suppose).

File: test/solvers/functional/functional_test.py (hunk 1 above)
Comment (kohr-h) [inline test/solvers/functional/functional_test.py]: The thing is that with the default convex conjugate added, this line won't raise any more but rather return an object that cannot be evaluated, but still asked for its convex conjugate and proximal.

**Comment by kohr-h** [comment]
> Looks very good to me! :)

Same here.

File: test/solvers/functional/functional_test.py (hunk 1 above)
Comment (adler-j) [inline test/solvers/functional/functional_test.py]: So this test should simply be changed to testing that we cannot evaluate it

**Comment by kohr-h** [comment]
@adler-j  Please write a release note on this.
